File: app/src/main/kotlin/br/com/sailboat/template/ui/base/BaseFragment.java
package br.com.sailboat.template.ui.base;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.CoordinatorLayout;
import android.support.design.widget.Snackbar;
import android.support.v4.app.Fragment;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import javax.inject.Inject;

import br.com.sailboat.template.R;
import br.com.sailboat.template.ui.dialog.MessageDialog;
import br.com.sailboat.template.ui.dialog.ProgressDialog;
import br.com.sailboat.template.ui.helper.Extras;
import br.com.sailboat.template.ui.helper.UIHelper;

public abstract class BaseFragment<P extends BaseMvpContract.Presenter> extends Fragment implements BaseMvpContract.View {

    @Inject
    P presenter;
    private ProgressDialog progressDialog;

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        inject();
        setHasOptionsMenu(true);
        setRetainInstance(true);
    }

    protected void inject() {
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(getLayoutId(), container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViews();
        initListeners();
        presenter.setView(this);
        presenter.onViewCreated();
    }

    protected void initViews() {
    }

    protected void initListeners() {
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        presenter.onDestroyView();
    }

    @Override
    public void showProgress() {
        progressDialog = new ProgressDialog();
        progressDialog.show(getFragmentManager(), "PROGRESS");
    }

    @Override
    public void hideProgress() {
        if (progressDialog != null) {
            progressDialog.dismiss();
            progressDialog = null;
        }
    }

    public void showErrorMessage(int msgId) {
        MessageDialog.Companion.show(getFragmentManager(), R.string.dlg_title_warning, msgId);
    }

    public void showFeedbackMessage(int msgId) {
        if (getView() instanceof CoordinatorLayout) {
            Snackbar.make(getView(), msgId, 4000).show();
        } else {
            Toast.makeText(getActivity(), msgId, Toast.LENGTH_LONG).show();
        }
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        switch (resultCode) {
            case Activity.RESULT_OK: {
                onResultOk(requestCode, data);
                break;
            }
            case Activity.RESULT_CANCELED: {
                onResultCanceled(requestCode, data);
                break;
            }
        }

        postResult(requestCode, data);
    }

    protected void onResultOk(int requestCode, Intent data) {
        if (Extras.INSTANCE.hasFeedbackMessage(data)) {
            showFeedbackMessage(Extras.INSTANCE.getFeedbackMessage(data));
        }
    }

    protected void onResultCanceled(int requestCode, Intent data) {
        if (Extras.INSTANCE.hasErrorMessage(data)) {
            showErrorMessage(Extras.INSTANCE.getErrorMessage(data));
        }
    }

    protected void postResult(int requestCode, Intent data) {
        presenter.postResult();
    }

    @Override
    public void closeKeyboard() {
        UIHelper.INSTANCE.closeKeyboard(getActivity());
    }

    @Override
    public void disableKeyboardOnStart() {
        UIHelper.INSTANCE.disableKeyboardOnStart(getActivity());
    }

    @Override
    public void closeWithSuccess() {
        getActivity().setResult(Activity.RESULT_OK);
        getActivity().finish();
    }

    public void closeWithSuccess(int msgId) {
        Intent intent = new Intent();
        Extras.INSTANCE.putFeedbackMessage(intent, msgId);

        getActivity().setResult(Activity.RESULT_OK, intent);
        getActivity().finish();
    }

    @Override
    public void closeWithFailure() {
        getActivity().setResult(Activity.RESULT_CANCELED);
        getActivity().finish();
    }

    public void closeWithFailure(int msgId) {
        Intent intent = new Intent();
        Extras.INSTANCE.putFeedbackMessage(intent, msgId);

        getActivity().setResult(Activity.RESULT_CANCELED);
        getActivity().finish();
    }

    @Override
    public void closeWithFailureDefaultMessage() {
        closeWithFailure(R.string.dlg_msg_error);
    }

    @Override
    public void logError(@NonNull Exception e) {
        String msg = "An error occurred while performing the operation";

        if (e.getMessage() != null) {
            msg = e.getMessage();
        }

        Log.e("APP_ERROR_LOG", msg, e);
    }

    protected abstract int getLayoutId();

    public P getPresenter() {
        return presenter;
    }

}

File: app/src/main/kotlin/br/com/sailboat/template/ui/base/BaseFragment.kt
package br.com.sailboat.template.ui.base;

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.design.widget.CoordinatorLayout
import android.support.design.widget.Snackbar
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import br.com.sailboat.template.R
import br.com.sailboat.template.ui.dialog.MessageDialog
import br.com.sailboat.template.ui.dialog.ProgressDialog
import br.com.sailboat.template.ui.helper.Extras
import br.com.sailboat.template.ui.helper.UIHelper
import javax.inject.Inject

abstract class BaseFragment<P : BaseMvpContract.Presenter> : Fragment(), BaseMvpContract.View {

    @Inject
    lateinit var presenter: P
    private var progressDialog: ProgressDialog? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        inject()
        setHasOptionsMenu(true)
        retainInstance = true
    }

    open fun inject() {}

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(getLayoutId(), container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
        initListeners()
        presenter.setView(this)
        presenter.onViewCreated()
    }

    open fun initViews() {}

    open fun initListeners() {}

    override fun onDestroyView() {
        super.onDestroyView()
        presenter.onDestroyView()
    }

    override fun showProgress() {
        progressDialog = ProgressDialog()
        progressDialog?.show(fragmentManager!!, "PROGRESS")
    }

    override fun hideProgress() {
        if (progressDialog != null) {
            progressDialog?.dismiss()
            progressDialog = null
        }
    }

    fun showErrorMessage(msgId: Int) {
        MessageDialog.show(fragmentManager!!, R.string.dlg_title_warning, msgId)
    }

    fun showFeedbackMessage(msgId: Int) {
        if (view is CoordinatorLayout) {
            Snackbar.make(view!!, msgId, 4000).show()
        } else {
            Toast.makeText(activity, msgId, Toast.LENGTH_LONG).show()
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        when (resultCode) {
            Activity.RESULT_OK -> {
                onResultOk(requestCode, data)
            }
            Activity.RESULT_CANCELED -> {
                onResultCanceled(requestCode, data)
            }
        }

        postResult(requestCode, data)
    }

    open fun onResultOk(requestCode: Int, data: Intent?) {
        if (Extras.hasFeedbackMessage(data)) {
            showFeedbackMessage(Extras.getFeedbackMessage(data)!!)
        }
    }

    open fun onResultCanceled(requestCode: Int, data: Intent?) {
        if (Extras.hasErrorMessage(data)) {
            showErrorMessage(Extras.getErrorMessage(data)!!)
        }
    }

    open fun postResult(requestCode: Int, data: Intent?) {
        presenter.postResult()
    }

    override fun closeKeyboard() {
        UIHelper.closeKeyboard(activity)
    }

    override fun disableKeyboardOnStart() {
        UIHelper.disableKeyboardOnStart(activity)
    }

    override fun closeWithSuccess() {
        activity?.setResult(Activity.RESULT_OK)
        activity?.finish()
    }

    fun closeWithSuccess(msgId: Int) {
        val intent = Intent()
        Extras.putFeedbackMessage(intent, msgId)

        activity?.setResult(Activity.RESULT_OK, intent)
        activity?.finish()
    }

    override fun closeWithFailure() {
        activity?.setResult(Activity.RESULT_CANCELED)
        activity?.finish()
    }

    fun closeWithFailure(msgId: Int) {
        val intent = Intent()
        Extras.putFeedbackMessage(intent, msgId)

        activity?.setResult(Activity.RESULT_CANCELED)
        activity?.finish()
    }

    override fun closeWithFailureDefaultMessage() {
        closeWithFailure(R.string.dlg_msg_error)
    }

    override fun logError(e: Exception) {
        var msg = "An error occurred while performing the operation"

        if (e.message != null) {
            msg = e.message!!
        }

        Log.e("APP_ERROR_LOG", msg, e)
    }

    protected abstract fun getLayoutId(): Int

}